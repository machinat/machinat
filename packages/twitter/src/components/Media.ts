import { NativeElement } from '@sociably/core';
import { makeUnitSegment, UnitSegment } from '@sociably/core/renderer';
import makeTwitterComponent from '../utils/makeTwitterComponent';
import { TwitterSegmentValue, TwitterComponent, MediaType } from '../types';

/**
 * @category Props
 */
export type MediaProps = {
  /** Set to true if media asset will be reused for multiple Direct Messages. Default is false */
  shared?: boolean;
  /** The tag to store media asset with */
  assetTag?: string;
  /** A string enum value which identifies a media usecase. This identifier is used to enforce usecase specific constraints (e.g. file size, video duration) and enable advanced features */
  mediaCategory?: string;
  /** A comma-separated list of user IDs to set as additional owners allowed to use the returned media_id in Tweets or Cards. Up to 100 additional owners may be specified */
  additionalOwners?: string[];
  /** The media file url */
  url?: string;
  /** The uploaded media id */
  mediaId?: string;
  /** The file content to be upload */
  fileData?: Buffer | NodeJS.ReadableStream;
  /** The content type of the file. Required when uploading using `file` */
  fileType?: string;
  /** The file size in bytes. Required when uploading using `file` */
  fileSize?: number;
};

const makeMediaComponent = (name: string, mediaType: MediaType) =>
  makeTwitterComponent(
    {
      [name](
        node: NativeElement<MediaProps, TwitterComponent<MediaProps>>,
        path: string
      ): UnitSegment<TwitterSegmentValue>[] {
        const {
          shared,
          assetTag,
          url,
          mediaId,
          fileData,
          fileSize,
          fileType,
          mediaCategory,
          additionalOwners,
        } = node.props;

        if (
          !(url || mediaId || fileData) ||
          (url && mediaId) ||
          (url && fileData) ||
          (fileData && mediaId)
        ) {
          throw new TypeError(
            'there should be exactly one of "url", "mediaId" or "fileData" prop'
          );
        }
        if (fileData && (!fileType || !fileSize)) {
          throw new TypeError(
            '"fileType" and "fileSize" props are required when using "fileData" source'
          );
        }

        const params = {
          shared: shared ? 'true' : undefined,
          media_category: mediaCategory,
          additional_owners: additionalOwners
            ? additionalOwners.join(',')
            : undefined,
        };

        return [
          makeUnitSegment(node, path, {
            type: 'media',
            attachment: mediaId
              ? {
                  type: mediaType,
                  source: { type: 'id', id: mediaId, params },
                }
              : url
              ? {
                  type: mediaType,
                  source: { type: 'url', url, params },
                }
              : fileData
              ? {
                  type: mediaType,
                  source: {
                    type: 'file',
                    params: {
                      ...params,
                      total_bytes: fileSize,
                      media_type: fileType,
                    },
                    fileData,
                    assetTag,
                  },
                }
              : (null as never),
          }),
        ];
      },
    }[name]
  );

/**
 * Send a photo by direct message or attach it to a tweet
 * @category Component
 * @props {@link MediaProps}
 * @guides Check official [guide](https://developer.twitter.com/en/docs/twitter-api/v1/media/upload-media/overview).
 */
export const Photo: TwitterComponent<
  MediaProps,
  UnitSegment<TwitterSegmentValue>
> = makeMediaComponent('Photo', 'photo');

/**
 * Send a video by direct message or attach it to a tweet
 * @category Component
 * @props {@link MediaProps}
 * @guides Check official [guide](https://developer.twitter.com/en/docs/twitter-api/v1/media/upload-media/overview).
 */
export const Video: TwitterComponent<
  MediaProps,
  UnitSegment<TwitterSegmentValue>
> = makeMediaComponent('Video', 'video');

/**
 * Send an animated gif by direct message or attach it to a tweet
 * @category Component
 * @props {@link MediaProps}
 * @guides Check official [guide](https://developer.twitter.com/en/docs/twitter-api/v1/media/upload-media/overview).
 */
export const AnimatedGif: TwitterComponent<
  MediaProps,
  UnitSegment<TwitterSegmentValue>
> = makeMediaComponent('AnimatedGif', 'animated_gif');
