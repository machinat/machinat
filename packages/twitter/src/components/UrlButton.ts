import { makePartSegment, PartSegment } from '@sociably/core/renderer';
import makeTwitterComponent from '../utils/makeTwitterComponent.js';
import { TwitterComponent } from '../types.js';

/**
 * @category Props
 */
export type UrlButtonProps = {
  /** The text that will be displayed to the user on each button. Max string length of 36 characters */
  label: string;
  /** A valid http or https target URL of the button */
  url: string;
};

/**
 * Attach a button for opening a URL to the direct message
 * @category Component
 * @props {@link UrlButtonProps}
 * @guides Check official [reference](https://developer.twitter.com/en/docs/twitter-api/v1/direct-messages/buttons/api-reference/buttons).
 */
export const UrlButton: TwitterComponent<
  UrlButtonProps,
  PartSegment<unknown>
> = makeTwitterComponent(function UrlButton(node, path) {
  const { label, url } = node.props;
  return [makePartSegment(node, path, { type: 'web_url', label, url })];
});
