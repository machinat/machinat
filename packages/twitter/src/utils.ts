import { annotateNativeComponent } from '@machinat/core/renderer';
import TwitterUser from './User';
import TwitterChat from './Chat';
import TweetTarget from './TweetTarget';
import { TWITTER } from './constant';
import type {
  RawEntities,
  UrlEntity,
  MentionEntity,
  RawMedia,
  Media,
  MediaSource,
  DirectMessageSegmentValue,
  TweetSegmentValue,
} from './types';

export const makeTwitterComponent = annotateNativeComponent(TWITTER);

export const polishMentionEntity = (
  raw: RawEntities['user_mentions'][number]
): MentionEntity => ({
  user: new TwitterUser(raw.id_str),
  id: raw.id_str,
  name: raw.name,
  screenName: raw.screen_name,
  indices: raw.indices,
});

export const polishUrlEntity = (
  raw: RawEntities['urls'][number]
): UrlEntity => ({
  url: raw.url,
  displayUrl: raw.display_url,
  expandedUrl: raw.expanded_url,
  indices: raw.indices,
});

export const polishMedia = (raw: RawMedia): Media => {
  const videoInfo = raw.video_info;
  return {
    id: raw.id_str,
    type: raw.type,
    inlineUrl: raw.url,
    displayUrl: raw.display_url,
    expandedUrl: raw.expanded_url,
    indices: raw.indices,
    downloadUrl: raw.media_url_https,
    sizes: raw.sizes,
    sourceStatusId: raw.source_status_id_str,
    videoInfo: videoInfo
      ? {
          aspectRatio: videoInfo.aspect_ratio,
          durationMs: videoInfo.duration_millis,
          variants: videoInfo.variants.map((rawVariant) => ({
            url: rawVariant.url,
            bitrate: rawVariant.bitrate,
            contentType: rawVariant.content_type,
          })),
        }
      : null,
    additionalMediaInfo: raw.additional_media_info || null,
  } as Media;
};

export const annotateDmMedia = (media: MediaSource): MediaSource =>
  media.sourceType === 'id'
    ? media
    : {
        ...media,
        parameters: {
          ...media.parameters,
          media_category:
            media.parameters.media_category ||
            (media.type === 'photo'
              ? 'dm_image'
              : media.type === 'video'
              ? 'dm_video'
              : media.type === 'animated_gif'
              ? 'dm_gif'
              : undefined),
        },
      };

export const annotateTweetMedia = (media: MediaSource): MediaSource =>
  media.sourceType === 'id'
    ? media
    : {
        ...media,
        parameters: {
          ...media.parameters,
          media_category:
            media.parameters.media_category ||
            (media.type === 'photo'
              ? 'tweet_image'
              : media.type === 'video'
              ? 'tweet_video'
              : media.type === 'animated_gif'
              ? 'tweet_gif'
              : undefined),
        },
      };

export const createDmSegmentValue = (
  text?: string,
  media?: MediaSource
): DirectMessageSegmentValue => ({
  type: 'dm',
  request: {
    method: 'POST',
    href: '1.1/direct_messages/events/new.json',
    parameters: {
      event: {
        type: 'message_create',
        message_create: {
          target: { recipient_id: '' },
          message_data: {
            text,
            attachment: media
              ? {
                  type: 'media',
                  media: { id: '' },
                }
              : undefined,
          },
        },
      },
    },
  },
  accomplishRequest: (target: TwitterChat, request, mediaResults) => {
    // eslint-disable-next-line no-param-reassign
    request.parameters.event.message_create.target.recipient_id = target.id;

    if (mediaResults) {
      // eslint-disable-next-line no-param-reassign, prefer-destructuring
      request.parameters.event.message_create.message_data.attachment.media.id =
        mediaResults[0];
    }
    return request;
  },
  mediaSources: media ? [annotateDmMedia(media)] : null,
});

export const createTweetSegmentValue = (
  parameters: Record<string, any>,
  mediaSources?: null | MediaSource[],
  dmLink?:
    | boolean
    | string
    | { recipientId?: string; text?: string; welcomeMessageId?: string }
): TweetSegmentValue => ({
  type: 'tweet',
  request: {
    method: 'POST',
    href: '2/tweets',
    parameters,
  },
  mediaSources: mediaSources ? mediaSources.map(annotateTweetMedia) : null,
  accomplishRequest: (target: TweetTarget, request, mediaIds) => {
    const params = request.parameters;
    return {
      ...request,
      parameters: {
        ...request.parameters,
        reply: target.tweetId
          ? {
              in_reply_to_tweet_id: target.tweetId,
              exclude_reply_user_ids: params.reply?.exclude_reply_user_ids,
            }
          : undefined,
        media: mediaIds
          ? {
              media_ids: mediaIds,
              tagged_user_ids: params.media?.tagged_user_ids,
            }
          : undefined,
        direct_message_deep_link:
          typeof dmLink === 'string'
            ? dmLink
            : typeof dmLink === 'object'
            ? `https://twitter.com/messages/compose?recipient_id=${
                dmLink.recipientId || target.agentId
              }${
                dmLink.welcomeMessageId
                  ? `&welcome_message_id=${dmLink.welcomeMessageId}`
                  : ''
              }${dmLink.text ? `&text=${encodeURIComponent(dmLink.text)}` : ''}`
            : dmLink
            ? `https://twitter.com/messages/compose?recipient_id=${target.agentId}`
            : undefined,
      },
    };
  },
});

let lastTime = 0;
let count = 0;
export const getTimeId = (): string => {
  const now = Date.now();
  if (lastTime === now) {
    count += 1;
  } else {
    lastTime = now;
    count = 0;
  }
  return `${now}-${count}`;
};
