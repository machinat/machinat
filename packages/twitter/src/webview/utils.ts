import type { ContextDetails } from '@sociably/auth';
import TwitterChat from '../Chat';
import TwitterUser from '../User';
import UserProfile from '../UserProfile';
import type { TwitterAuthContext, TwitterAuthData } from './types';

// eslint-disable-next-line import/prefer-default-export
export const getAuthContextDetails = ({
  agent: agentId,
  user: { id: userId, data: rawUserData },
}: TwitterAuthData): ContextDetails<TwitterAuthContext> => {
  return {
    agentId,
    channel: new TwitterUser(agentId),
    user: new TwitterUser(userId, rawUserData),
    thread: new TwitterChat(agentId, userId),
    userProfile: new UserProfile(rawUserData),
  };
};
