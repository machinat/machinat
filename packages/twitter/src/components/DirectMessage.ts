import { MachinatNode } from '@machinat/core';
import { makeUnitSegment, UnitSegment } from '@machinat/core/renderer';
import { formatNode } from '@machinat/core/utils';
import makeTwitterComponent from '../utils/makeTwitterComponent';
import createDmSegmentValue from '../utils/createDmSegmentValue';
import { TwitterSegmentValue, TwitterComponent } from '../types';

/**
 * @category Props
 */
export type DirectMessageProps = {
  /** Texual content of the direct message */
  children?: MachinatNode;
  /** The media attached to the message. Should contain exactly one {@link Photo}, {@link Video} or {@link AnimatedGif} */
  media?: MachinatNode;
  /** The twitter place represent the location. You can search places using [`geo/search API`](https://developer.twitter.com/en/docs/twitter-api/v1/geo/places-near-location/api-reference/get-geo-search) */
  placeId?: string;
  /** The coordinates of the location */
  coordinates?: { latitude: number; longitude: number };
  /** Url buttons to be attached below the messages. Should contain only {@link UrlButton} */
  buttons?: MachinatNode;
  /** Quick replies to be attached after the messages. Should contain only {@link QuickReply} */
  quickReplies?: MachinatNode;
  /** The custome profile to send the messages with */
  customProfileId?: string;
};

/**
 * Send a direct message
 * @category Component
 * @props {@link DirectMessageProps}
 * @guides Check official [guide](https://developer.twitter.com/en/docs/twitter-api/v1/direct-messages/message-attachments/guides/attaching-location).
 */
export const DirectMessage: TwitterComponent<
  DirectMessageProps,
  UnitSegment<TwitterSegmentValue>
> = makeTwitterComponent(async function DirectMessage(node, path, render) {
  const {
    children,
    media,
    buttons,
    quickReplies,
    placeId,
    coordinates,
    customProfileId,
  } = node.props;

  if (
    (media && placeId) ||
    (media && coordinates) ||
    (placeId && coordinates)
  ) {
    throw new TypeError(
      'there should be exactly one of "media", "placeId" or "coordinates" prop'
    );
  }

  const [
    contentSegments,
    mediaSegments,
    buttonsSegments,
    quickRepliesSegments,
  ] = await Promise.all([
    render<TwitterSegmentValue>(children, '.children'),
    render<TwitterSegmentValue>(media, '.media'),
    render(buttons, '.buttons'),
    render(quickReplies, '.quickReplies'),
  ]);

  if (!contentSegments && !mediaSegments && !placeId && !coordinates) {
    throw new TypeError(`no text or attachment in <DirectMessage/>`);
  }

  for (const segment of contentSegments || []) {
    if (segment.type !== 'text') {
      throw new TypeError(
        `non-texual element ${formatNode(
          segment.node
        )} can't be placed under <DirectMessage/>`
      );
    }
  }

  if (
    mediaSegments &&
    (mediaSegments.length > 1 ||
      (mediaSegments[0].type !== 'unit' && mediaSegments[0].type !== 'raw') ||
      mediaSegments[0].value.type !== 'media')
  ) {
    throw new TypeError(
      `"media" prop should contain only 1 "Photo", "Video" or "AnimatedGif"`
    );
  }

  const dmSegmentValue = createDmSegmentValue(
    contentSegments?.[0].value,
    mediaSegments?.[0].value.media
  );

  const messageCreateParams =
    dmSegmentValue.request.parameters.event.message_create;
  if (placeId) {
    messageCreateParams.message_data.attachment = {
      type: 'location',
      location: {
        type: 'shared_place',
        shared_place: {
          place: { id: placeId },
        },
      },
    };
  }
  if (coordinates) {
    messageCreateParams.message_data.attachment = {
      type: 'location',
      location: {
        type: 'shared_coordinate',
        shared_coordinate: {
          coordinates: {
            type: 'Point',
            coordinates: [coordinates.longitude, coordinates.latitude],
          },
        },
      },
    };
  }
  if (customProfileId) {
    messageCreateParams.custom_profile_id = customProfileId;
  }
  if (quickRepliesSegments) {
    messageCreateParams.message_data.quick_reply = {
      type: 'options',
      options: quickRepliesSegments.map(({ value }) => value),
    };
  }
  if (buttonsSegments) {
    messageCreateParams.message_data.ctas = buttonsSegments.map(
      ({ value }) => value
    );
  }

  return [makeUnitSegment(node, path, dmSegmentValue)];
});
