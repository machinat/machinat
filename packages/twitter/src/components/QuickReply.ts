import { makePartSegment, PartSegment } from '@machinat/core/renderer';
import { makeTwitterComponent } from '../utils';
import { TwitterComponent } from '../types';

/**
 * @category Props
 */
export type QuickReplyProps = {
  /** The text label displayed on the button face. Label text is returned as the user's message response. String, max length of 36 characters including spaces. Values with URLs are not allowed and will return an error */
  label: string;
  /** Optional description text displayed under label text. All options must have this property defined if property is present in any option. Text is auto-wrapped and will display on a max of two lines and supports n for controling line breaks. Description text is not include in the user's message response. String, max length of 72 characters including spaces */
  description?: string;
  /** Metadata that will be sent back in the webhook request. String, max length of 1,000 characters including spaces */
  metadata?: string;
};

/**
 * QuickReply a tweet
 * @category Component
 * @props {@link QuickReplyProps}
 * @guides Check official [guide](https://developer.twitter.com/en/docs/twitter-api/v1/direct-messages/quick-replies/overview).
 */
export const QuickReply: TwitterComponent<
  QuickReplyProps,
  PartSegment<unknown>
> = makeTwitterComponent(function QuickReply(node, path) {
  const { label, description, metadata } = node.props;

  return [makePartSegment(node, path, { label, description, metadata })];
});
