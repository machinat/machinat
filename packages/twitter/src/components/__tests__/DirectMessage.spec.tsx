import Sociably from '@sociably/core';
import { isNativeType } from '@sociably/core/utils';
import Renderer from '@sociably/core/renderer';
import TwitterChat from '../../Chat';
import { Photo } from '../Media';
import { UrlButton } from '../UrlButton';
import { QuickReply } from '../QuickReply';
import { DirectMessage } from '../DirectMessage';

const renderer = new Renderer('twitter', async (node, path) => [
  { type: 'text', path, node, value: node.type },
]);
const render = (element) => renderer.render(element, null, null);

it('is a valid Component', () => {
  expect(typeof DirectMessage).toBe('function');
  expect(isNativeType(<DirectMessage placeId="12345" />)).toBe(true);
  expect(DirectMessage.$$platform).toBe('twitter');
});

test('rendering with plain text', async () => {
  let segments = await render(<DirectMessage>Hello World</DirectMessage>);
  expect(segments).toMatchSnapshot();
  let { request, accomplishRequest } = (segments as any)[0].value;
  expect(accomplishRequest(new TwitterChat('12345', '67890'), request, null))
    .toMatchInlineSnapshot(`
    Object {
      "href": "1.1/direct_messages/events/new.json",
      "method": "POST",
      "params": Object {
        "event": Object {
          "message_create": Object {
            "message_data": Object {
              "attachment": undefined,
              "text": "Hello World",
            },
            "target": Object {
              "recipient_id": "67890",
            },
          },
          "type": "message_create",
        },
      },
    }
  `);
  segments = await render(
    <DirectMessage customProfileId="11111">
      <foo />
    </DirectMessage>
  );
  expect(segments).toMatchSnapshot();
  ({ request, accomplishRequest } = (segments as any)[0].value);
  expect(accomplishRequest(new TwitterChat('12345', '67890'), request, null))
    .toMatchInlineSnapshot(`
    Object {
      "href": "1.1/direct_messages/events/new.json",
      "method": "POST",
      "params": Object {
        "event": Object {
          "message_create": Object {
            "custom_profile_id": "11111",
            "message_data": Object {
              "attachment": undefined,
              "text": "foo",
            },
            "target": Object {
              "recipient_id": "67890",
            },
          },
          "type": "message_create",
        },
      },
    }
  `);
});

test('rendering with media', async () => {
  const segments = await render(
    <DirectMessage media={<Photo mediaId="11111" />} />
  );
  expect(segments).toMatchSnapshot();
  const { request, accomplishRequest } = (segments as any)[0].value;
  expect(
    accomplishRequest(new TwitterChat('12345', '67890'), request, ['11111'])
  ).toMatchInlineSnapshot(`
    Object {
      "href": "1.1/direct_messages/events/new.json",
      "method": "POST",
      "params": Object {
        "event": Object {
          "message_create": Object {
            "message_data": Object {
              "attachment": Object {
                "media": Object {
                  "id": "11111",
                },
                "type": "media",
              },
              "text": undefined,
            },
            "target": Object {
              "recipient_id": "67890",
            },
          },
          "type": "message_create",
        },
      },
    }
  `);
});

test('rendering with buttons', async () => {
  const segments = await render(
    <DirectMessage
      buttons={
        <>
          <UrlButton label="foo" url="http://sociably.io/foo" />
          <UrlButton label="bar" url="http://sociably.io/bar" />
          <UrlButton label="baz" url="http://sociably.io/baz" />
        </>
      }
    >
      Guess one
    </DirectMessage>
  );
  expect(segments).toMatchSnapshot();
  const { request, accomplishRequest } = (segments as any)[0].value;
  expect(accomplishRequest(new TwitterChat('12345', '67890'), request, null))
    .toMatchInlineSnapshot(`
    Object {
      "href": "1.1/direct_messages/events/new.json",
      "method": "POST",
      "params": Object {
        "event": Object {
          "message_create": Object {
            "message_data": Object {
              "attachment": undefined,
              "ctas": Array [
                Object {
                  "label": "foo",
                  "type": "web_url",
                  "url": "http://sociably.io/foo",
                },
                Object {
                  "label": "bar",
                  "type": "web_url",
                  "url": "http://sociably.io/bar",
                },
                Object {
                  "label": "baz",
                  "type": "web_url",
                  "url": "http://sociably.io/baz",
                },
              ],
              "text": "Guess one",
            },
            "target": Object {
              "recipient_id": "67890",
            },
          },
          "type": "message_create",
        },
      },
    }
  `);
});

test('rendering with quick replies', async () => {
  const segments = await render(
    <DirectMessage
      quickReplies={
        <>
          <QuickReply label="foo" description="FOOOOOOO" metadata="FOO" />
          <QuickReply label="bar" description="BAAAAAAR" metadata="BAR" />
          <QuickReply label="baz" description="BAAAAAAZ" metadata="BAZ" />
        </>
      }
    >
      Guess one
    </DirectMessage>
  );
  expect(segments).toMatchSnapshot();
  const { request, accomplishRequest } = (segments as any)[0].value;
  expect(accomplishRequest(new TwitterChat('12345', '67890'), request, null))
    .toMatchInlineSnapshot(`
    Object {
      "href": "1.1/direct_messages/events/new.json",
      "method": "POST",
      "params": Object {
        "event": Object {
          "message_create": Object {
            "message_data": Object {
              "attachment": undefined,
              "quick_reply": Object {
                "options": Array [
                  Object {
                    "description": "FOOOOOOO",
                    "label": "foo",
                    "metadata": "FOO",
                  },
                  Object {
                    "description": "BAAAAAAR",
                    "label": "bar",
                    "metadata": "BAR",
                  },
                  Object {
                    "description": "BAAAAAAZ",
                    "label": "baz",
                    "metadata": "BAZ",
                  },
                ],
                "type": "options",
              },
              "text": "Guess one",
            },
            "target": Object {
              "recipient_id": "67890",
            },
          },
          "type": "message_create",
        },
      },
    }
  `);
});

test('rendering with placeId', async () => {
  const segments = await render(
    <DirectMessage placeId="12345">Hello there</DirectMessage>
  );
  expect(segments).toMatchSnapshot();
  const { request, accomplishRequest } = (segments as any)[0].value;
  expect(accomplishRequest(new TwitterChat('12345', '67890'), request, null))
    .toMatchInlineSnapshot(`
    Object {
      "href": "1.1/direct_messages/events/new.json",
      "method": "POST",
      "params": Object {
        "event": Object {
          "message_create": Object {
            "message_data": Object {
              "attachment": Object {
                "location": Object {
                  "shared_place": Object {
                    "place": Object {
                      "id": "12345",
                    },
                  },
                  "type": "shared_place",
                },
                "type": "location",
              },
              "text": "Hello there",
            },
            "target": Object {
              "recipient_id": "67890",
            },
          },
          "type": "message_create",
        },
      },
    }
  `);
});

test('rendering with coordinates', async () => {
  const segments = await render(
    <DirectMessage coordinates={{ latitude: 25.1, longitude: 121.6 }} />
  );
  expect(segments).toMatchSnapshot();
  const { request, accomplishRequest } = (segments as any)[0].value;
  expect(accomplishRequest(new TwitterChat('12345', '67890'), request, null))
    .toMatchInlineSnapshot(`
    Object {
      "href": "1.1/direct_messages/events/new.json",
      "method": "POST",
      "params": Object {
        "event": Object {
          "message_create": Object {
            "message_data": Object {
              "attachment": Object {
                "location": Object {
                  "shared_coordinate": Object {
                    "coordinates": Object {
                      "coordinates": Array [
                        121.6,
                        25.1,
                      ],
                      "type": "Point",
                    },
                  },
                  "type": "shared_coordinate",
                },
                "type": "location",
              },
              "text": undefined,
            },
            "target": Object {
              "recipient_id": "67890",
            },
          },
          "type": "message_create",
        },
      },
    }
  `);
});

it('throw if no content and attachment available', async () => {
  await expect(
    render(<DirectMessage />)
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"no text or attachment in <DirectMessage/>"`
  );
});

it('throw if "media", "placeId" and "coordinates" are duplicated', async () => {
  await expect(
    render(
      <DirectMessage
        placeId="12345"
        coordinates={{ latitude: 25.1, longitude: 121.6 }}
      />
    )
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"there should be exactly one of \\"media\\", \\"placeId\\" or \\"coordinates\\" prop"`
  );
  await expect(
    render(<DirectMessage placeId="12345" media={<Photo mediaId="11111" />} />)
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"there should be exactly one of \\"media\\", \\"placeId\\" or \\"coordinates\\" prop"`
  );
});

it('throw if invalid content received', async () => {
  await expect(
    render(
      <DirectMessage>
        <Photo mediaId="11111" />
      </DirectMessage>
    )
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"non-texual element <Photo /> can't be placed under <DirectMessage/>"`
  );
  await expect(
    render(
      <DirectMessage>
        <Sociably.Pause />
      </DirectMessage>
    )
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"non-texual element <Pause /> can't be placed under <DirectMessage/>"`
  );
});

it('throw if invalid media received', async () => {
  await expect(
    render(<DirectMessage media="foo" />)
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"\\"media\\" prop should contain only 1 \\"Photo\\", \\"Video\\" or \\"AnimatedGif\\""`
  );
  await expect(
    render(<DirectMessage media={<Sociably.Pause />} />)
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"\\"media\\" prop should contain only 1 \\"Photo\\", \\"Video\\" or \\"AnimatedGif\\""`
  );
});

it('throw if more than 1 media', async () => {
  await expect(
    render(
      <DirectMessage
        media={
          <>
            <Photo mediaId="11111" />
            <Photo mediaId="22222" />
          </>
        }
      />
    )
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"\\"media\\" prop should contain only 1 \\"Photo\\", \\"Video\\" or \\"AnimatedGif\\""`
  );
});
