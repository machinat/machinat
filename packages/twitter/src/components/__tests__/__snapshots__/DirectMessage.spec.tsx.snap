// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering with buttons 1`] = `
Array [
  Object {
    "node": <DirectMessage
      buttons={
        <Sociably.Fragment>
          <UrlButton
            label="foo"
            url="http://sociably.io/foo"
          />
          <UrlButton
            label="bar"
            url="http://sociably.io/bar"
          />
          <UrlButton
            label="baz"
            url="http://sociably.io/baz"
          />
        </Sociably.Fragment>
      }
    >
      Guess one
    </DirectMessage>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": null,
      "request": Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "message_data": Object {
                "attachment": undefined,
                "ctas": Array [
                  Object {
                    "label": "foo",
                    "type": "web_url",
                    "url": "http://sociably.io/foo",
                  },
                  Object {
                    "label": "bar",
                    "type": "web_url",
                    "url": "http://sociably.io/bar",
                  },
                  Object {
                    "label": "baz",
                    "type": "web_url",
                    "url": "http://sociably.io/baz",
                  },
                ],
                "text": "Guess one",
              },
              "target": Object {
                "recipient_id": "",
              },
            },
            "type": "message_create",
          },
        },
      },
      "type": "dm",
    },
  },
]
`;

exports[`rendering with coordinates 1`] = `
Array [
  Object {
    "node": <DirectMessage
      coordinates={
        Object {
          "latitude": 25.1,
          "longitude": 121.6,
        }
      }
    />,
    "path": "$",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": null,
      "request": Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "message_data": Object {
                "attachment": Object {
                  "location": Object {
                    "shared_coordinate": Object {
                      "coordinates": Object {
                        "coordinates": Array [
                          121.6,
                          25.1,
                        ],
                        "type": "Point",
                      },
                    },
                    "type": "shared_coordinate",
                  },
                  "type": "location",
                },
                "text": undefined,
              },
              "target": Object {
                "recipient_id": "",
              },
            },
            "type": "message_create",
          },
        },
      },
      "type": "dm",
    },
  },
]
`;

exports[`rendering with media 1`] = `
Array [
  Object {
    "node": <DirectMessage
      media={
        <Photo
          mediaId="11111"
        />
      }
    />,
    "path": "$",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": Array [
        Object {
          "id": "11111",
          "parameters": Object {
            "additional_owners": undefined,
            "media_category": undefined,
            "shared": undefined,
          },
          "type": "id",
        },
      ],
      "request": Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "message_data": Object {
                "attachment": Object {
                  "media": Object {
                    "id": "",
                  },
                  "type": "media",
                },
                "text": undefined,
              },
              "target": Object {
                "recipient_id": "",
              },
            },
            "type": "message_create",
          },
        },
      },
      "type": "dm",
    },
  },
]
`;

exports[`rendering with placeId 1`] = `
Array [
  Object {
    "node": <DirectMessage
      placeId="12345"
    >
      Hello there
    </DirectMessage>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": null,
      "request": Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "message_data": Object {
                "attachment": Object {
                  "location": Object {
                    "shared_place": Object {
                      "place": Object {
                        "id": "12345",
                      },
                    },
                    "type": "shared_place",
                  },
                  "type": "location",
                },
                "text": "Hello there",
              },
              "target": Object {
                "recipient_id": "",
              },
            },
            "type": "message_create",
          },
        },
      },
      "type": "dm",
    },
  },
]
`;

exports[`rendering with plain text 1`] = `
Array [
  Object {
    "node": <DirectMessage>
      Hello World
    </DirectMessage>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": null,
      "request": Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "message_data": Object {
                "attachment": undefined,
                "text": "Hello World",
              },
              "target": Object {
                "recipient_id": "",
              },
            },
            "type": "message_create",
          },
        },
      },
      "type": "dm",
    },
  },
]
`;

exports[`rendering with plain text 3`] = `
Array [
  Object {
    "node": <DirectMessage
      customProfileId="11111"
    >
      <foo />
    </DirectMessage>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": null,
      "request": Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "custom_profile_id": "11111",
              "message_data": Object {
                "attachment": undefined,
                "text": "foo",
              },
              "target": Object {
                "recipient_id": "",
              },
            },
            "type": "message_create",
          },
        },
      },
      "type": "dm",
    },
  },
]
`;

exports[`rendering with quick replies 1`] = `
Array [
  Object {
    "node": <DirectMessage
      quickReplies={
        <Sociably.Fragment>
          <QuickReply
            description="FOOOOOOO"
            label="foo"
            metadata="FOO"
          />
          <QuickReply
            description="BAAAAAAR"
            label="bar"
            metadata="BAR"
          />
          <QuickReply
            description="BAAAAAAZ"
            label="baz"
            metadata="BAZ"
          />
        </Sociably.Fragment>
      }
    >
      Guess one
    </DirectMessage>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": null,
      "request": Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "message_data": Object {
                "attachment": undefined,
                "quick_reply": Object {
                  "options": Array [
                    Object {
                      "description": "FOOOOOOO",
                      "label": "foo",
                      "metadata": "FOO",
                    },
                    Object {
                      "description": "BAAAAAAR",
                      "label": "bar",
                      "metadata": "BAR",
                    },
                    Object {
                      "description": "BAAAAAAZ",
                      "label": "baz",
                      "metadata": "BAZ",
                    },
                  ],
                  "type": "options",
                },
                "text": "Guess one",
              },
              "target": Object {
                "recipient_id": "",
              },
            },
            "type": "message_create",
          },
        },
      },
      "type": "dm",
    },
  },
]
`;
